Open table api

Gr0d3N opened the pull request:
I've created 2 classes. OpenTable to make the API calls and Restaurant to create new restaurant objects.

Commit:
Add OpenTable class

Commit:
Add Restaurant class

AndreyPogorelyy commented on src/main/java/com/amazon/asksdk/cloudfoody/Restaurant.java:
```diff
@@ -0,0 +1,102 @@
+
+public class Restaurant extends OpenTable {
+
+    private String name, address, city, state, area, country, phone, reserveURL, mobileReserveURL, imageURL, postalCode;
```
NP: please split into several lines to improve readability.

Gr0d3N replied:
is it better to put every single variable on a separate line or just split them into 3 lines?

AndreyPogorelyy replied:
3 lines would be OK.

Gr0d3N replied:
Done!

AndreyPogorelyy commented on src/main/java/com/amazon/asksdk/cloudfoody/Restaurant.java:
```diff
@@ -0,0 +1,102 @@
+import java.util.ArrayList;
+import java.util.HashMap;
+
+public class Restaurant extends OpenTable {
```
Restaurant shouldn't extend OpenTable, because it is not a different version of "OpenTable", but a separate class.

Furthermore,  Restaurant class shouldn't have references to OpenTable methods. Please consider making Restaurant a POJO (Plain Old Java Object), and methods in OpenTable client returning instances of Restaurant. 

You could consider bringing a dependency on Jackson to map JSON responses into Restaurant instances (
http://tutorials.jenkov.com/java-json/jackson-objectmapper.html) 

https://github.com/FasterXML/jackson-core

Gr0d3N replied:
I made Restaurant a POJO.
I would like to use jackson to pam the JSON response to a restaurant class but there are many levels in the response.

AndreyPogorelyy commented on src/main/java/com/amazon/asksdk/cloudfoody/OpenTable.java:
```diff
@@ -0,0 +1,42 @@
+
+    private static final String BASE_URL = "http://opentable.herokuapp.com/api/";
+
+    /*
```
NP: Please change the comment formatting, because of the convention.

E.g. 
/**
 *
 */

Gr0d3N replied:
Done, please let me know if I understood this right

AndreyPogorelyy commented on src/main/java/com/amazon/asksdk/cloudfoody/OpenTable.java:
```diff
@@ -0,0 +1,42 @@
+        The method takes a restaurant name as an arguement and returns back a hash with
+        all the info regarding that restaurant
+     */
+    public static HashMap getRestaurantByName(String name) throws IOException {
```
Please return Restaurant instead of HashMap.

AndreyPogorelyy replied:
In addition, please consider improving the error handling in this class. In addition to IOException, please think of handling service response errors (4xx, 5xx). If possible, try to throw specific errors, so we can handle various scenarios. E.g. if the service returns 404, throw some "ObjectNotFound" error (please check if service response errors are already defined in some library, e.g. java.net).

AndreyPogorelyy commented on src/main/java/com/amazon/asksdk/cloudfoody/OpenTable.java:
```diff
@@ -0,0 +1,42 @@
+        The method takes a city name as an arguement and returns back a hash with
+        all the restaurants in that city and their info as well
+     */
+    public static HashMap getRestaurantsInCity(String city) throws IOException {
```
Please consider returning a collection of Restaurants instead.

AndreyPogorelyy commented on src/main/java/com/amazon/asksdk/cloudfoody/OpenTable.java:
```diff
@@ -0,0 +1,42 @@
+package com.amazon.asksdk.cloudfoody;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.*;
```
NP: It is the best practice to avoid import all (.*) syntax and import individual items.

Gr0d3N replied:
Done

Commit:
Split the Restaurant feilds into 3 lines to improve readablity

Commit:
Add a method to get all the cities in the api and inset them in a text file

Commit:
Move the logic of getting the restaurant from the Restaurant constructor to the OpenTable class

Commit:
Remove unnecessary import

Commit:
Delete the hello world sample

Commit:
Finalize the OpenTable client class
1. Modify the comments to match the convention
2. Modify the getRestaurantByName class to create a Restaurant object.
3. Modify the getRestaurantsIncity class to create an ArrayList of
Restarant objects.
4. Add methods that will generate text files with all the cities and the
restaurans in the OpenTableAPI.

Commit:
Modify the Restaurant class to be an POJO style class

Commit:
Change the package name

AndreyPogorelyy commented on src/main/java/com/amazon/awslabs/cloudfoody/OpenTable.java:
```diff
@@ -0,0 +1,145 @@
+    /**
+     *  The method writes the names of the restaurants for a given city to a text file
+     */
+    public static void writeRestaurantsNamesToFile(String city, String fileName) throws IOException {
```
Please remove methods which will not be used in the app.

AndreyPogorelyy commented on src/main/java/com/amazon/awslabs/cloudfoody/OpenTable.java:
```diff
@@ -0,0 +1,145 @@
+     *  The method will return the number of pages expected to ger from the API call GET /api/restaurants?city=Some+City
+     *  Note that number of pages was calculated assuming records per page will be set to maximum which is 100
+     */
+    public static int getNumberOfPages(String city) throws IOException {
```
Will we need this method in the skill?

AndreyPogorelyy commented on src/main/java/com/amazon/awslabs/cloudfoody/OpenTable.java:
```diff
@@ -0,0 +1,145 @@
+    /**
+     *  The method generates an ArrayList with all the available cities in the API
+     */
+    public static ArrayList getAllCities() throws IOException {
```
Would we need this method for the skill?

AndreyPogorelyy commented on src/main/java/com/amazon/awslabs/cloudfoody/OpenTable.java:
```diff
@@ -0,0 +1,145 @@
+     *  The method takes an ArrayList of cities and writes them to a text file
+     *  Each city will be in a separate line
+     */
+    public static void writeCitiesNamesToFile(String fileName) throws IOException{
```
Same question here.

NullGuru commented:
I am keeping them, so we can add more cities later on.

AndreyPogorelyy commented on src/main/java/com/amazon/awslabs/cloudfoody/OpenTable.java:
```diff
@@ -0,0 +1,145 @@
+     */
+    public static Restaurant getRestaurantByName(String name) throws IOException {
+        String fullURL = BASE_URL + "restaurants?name=" + name;
+        URLConnection connection = createConnection(fullURL);
```
Please try this:

        final ObjectMapper mapper = new ObjectMapper();
        final InputStreamReader streamReader = new InputStreamReader(connection.getInputStream());
        
        return mapper.readValue(streamReader, Restaurant.class);
